File: 2024/Day_6/Solution.go

package main

import (
    // "bufio"
    "fmt"
    "os"
    "strings"
	"path/filepath"
)

type Point struct {
    r, c int
}

type State struct {
    r, c int
    dir  string 
}

func createGrid(filename string) ([][]string, error) {
    cwd, err := os.Getwd()
    if err != nil {
        return nil, fmt.Errorf("error getting current directory: %v", err)
    }

    content, err := os.ReadFile(filepath.Join(cwd, "2024", "Day_6", "data.txt"))
    if err != nil {
        return nil, fmt.Errorf("error reading file: %v", err)
    }

    lines := strings.Split(string(content), "\n")
    grid := make([][]string, 0, len(lines))
    
    for _, line := range lines {
        if len(line) > 0 { 
            row := strings.Split(line, "")
            grid = append(grid, row)
        }
    }

    return grid, err
}

// func createGrid(lines []string) [][]string {
//     grid := make([][]string, len(lines))
//     for i, line := range lines {
//         grid[i] = strings.Split(line, "")
//     }
//     return grid
// }

func findStart(grid [][]string) Point {
    for i, row := range grid {
        for j, cell := range row {
            if cell == "^" {
                return Point{i, j}
            }
        }
    }
    return Point{0, 0}
}

func partOne(grid [][]string) int {
    dyDx := map[int]Point{
        0: {-1, 0},
        1: {0, 1},
        2: {1, 0},
        3: {0, -1},
    }
    directions := []string{"^", ">", "v", "<"}
    current := "^"
    pos := findStart(grid)
    visited := make(map[Point]bool)
    
    for {
        visited[pos] = true
        
        idx := indexOf(directions, current)
        delta := dyDx[idx]
        newR, newC := pos.r + delta.r, pos.c + delta.c
        
        if !inBounds(newR, newC, grid) {
            break
        }
        
        if grid[newR][newC] != "#" {
            pos.r, pos.c = newR, newC
            continue
        }
        
        current = directions[(idx + 1) % 4]
    }
    
    return len(visited)
}

func partTwo(grid [][]string) int {
    start := findStart(grid)
    cycles := 0
    
    for i := range grid {
        for j := range grid[0] {
            if grid[i][j] == "." && (i != start.r || j != start.c) {
                testGrid := copyGrid(grid)
                testGrid[i][j] = "#"
                if simulate(testGrid, start) {
                    cycles++
                }
            }
        }
    }
    return cycles
}

func simulate(grid [][]string, start Point) bool {
    dyDx := map[int]Point{
        0: {-1, 0},
        1: {0, 1},
        2: {1, 0},
        3: {0, -1},
    }
    directions := []string{"^", ">", "v", "<"}
    current := "^"
    pos := start
    visited := make(map[State]bool)
    
    for {
        state := State{pos.r, pos.c, current}
        if visited[state] {
            return true
        }
        visited[state] = true
        
        idx := indexOf(directions, current)
        delta := dyDx[idx]
        newR, newC := pos.r + delta.r, pos.c + delta.c
        
        if !inBounds(newR, newC, grid) {
            return false
        }
        
        if grid[newR][newC] != "#" {
            pos.r, pos.c = newR, newC
            continue
        }
        
        current = directions[(idx + 1) % 4]
    }
}

func copyGrid(grid [][]string) [][]string {
    newGrid := make([][]string, len(grid))
    for i := range grid {
        newGrid[i] = make([]string, len(grid[i]))
        copy(newGrid[i], grid[i])
    }
    return newGrid
}

func indexOf(slice []string, val string) int {
    for i, item := range slice {
        if item == val {
            return i
        }
    }
    return -1
}

func inBounds(r, c int, grid [][]string) bool {
    return r >= 0 && r < len(grid) && c >= 0 && c < len(grid[0])
}

func main() {
  
    grid, _ := createGrid("data.txt")
    fmt.Println("Part 1:", partOne(grid))
    fmt.Println("Part 2:", partTwo(grid))
}

File: 2024/Day_6/Solution.ts
import * as fs from 'fs';
import * as path from 'path';

interface Point {
    r: number;
    c: number;
}

interface State {
    r: number; 
    c: number;
    dir: string;
}


function createGrid(filename: string): string[][] {
    const content = fs.readFileSync(filename, 'utf8');
    return content.split('\n').map(row => row.split(''));
}

function findStart(grid: string[][]): Point {
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[i].length; j++) {
            if (grid[i][j] === '^') return {r: i, c: j};
        }
    }
    return {r: 0, c: 0};
}

function partOne(grid: string[][]): number {
    const dyDx: Record<number, Point> = {
        0: {r: -1, c: 0},
        1: {r: 0, c: 1},
        2: {r: 1, c: 0},
        3: {r: 0, c: -1}
    };
    const directions = ['^', '>', 'v', '<'];
    let current = '^';
    const pos = findStart(grid);
    const visited = new Set<string>();
    
    while (true) {
        visited.add(`${pos.r},${pos.c}`);
        
        const idx = directions.indexOf(current);
        const delta = dyDx[idx];
        const newR = pos.r + delta.r;
        const newC = pos.c + delta.c;
        
        if (!inBounds(newR, newC, grid)) break;
        
        if (grid[newR][newC] !== '#') {
            pos.r = newR;
            pos.c = newC;
            continue;
        }
        
        current = directions[(idx + 1) % 4];
    }
    
    return visited.size;
}

function partTwo(grid: string[][]): number {
    const start = findStart(grid);
    let cycles = 0;
    
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            if (grid[i][j] === '.' && (i !== start.r || j !== start.c)) {
                const testGrid = copyGrid(grid);
                testGrid[i][j] = '#';
                if (simulate(testGrid, start)) cycles++;
            }
        }
    }
    return cycles;
}

function simulate(grid: string[][], start: Point): boolean {
    const dyDx: Record<number, Point> = {
        0: {r: -1, c: 0},
        1: {r: 0, c: 1}, 
        2: {r: 1, c: 0},
        3: {r: 0, c: -1}
    };
    const directions = ['^', '>', 'v', '<'];
    let current = '^';
    const pos = {...start};
    const visited = new Set<string>();
    
    while (true) {
        const state = `${pos.r},${pos.c},${current}`;
        if (visited.has(state)) return true;
        visited.add(state);
        
        const idx = directions.indexOf(current);
        const delta = dyDx[idx];
        const newR = pos.r + delta.r;
        const newC = pos.c + delta.c;
        
        if (!inBounds(newR, newC, grid)) return false;
        
        if (grid[newR][newC] !== '#') {
            pos.r = newR;
            pos.c = newC;
            continue;
        }
        
        current = directions[(idx + 1) % 4];
    }
}

function copyGrid(grid: string[][]): string[][] {
    return grid.map(row => [...row]);
}

function inBounds(r: number, c: number, grid: string[][]): boolean {
    return r >= 0 && r < grid.length && c >= 0 && c < grid[0].length;
}

async function main() {
    const grid = await createGrid('2024/Day_6/data.txt');
    console.log('Part 1:', partOne(grid));
    console.log('Part 2:', partTwo(grid));
}

main();